src/components/views/settings/tabs/VoiceSettingsTab.js

```diff
@@ -0,0 +1,177 @@
+import AccessibleButton from "../../elements/AccessibleButton";
+import {SettingLevel} from "../../../../settings/SettingsStore";
+const Modal = require("../../../../Modal");
+const sdk = require("../../../../index");
+const MatrixClientPeg = require("../../../../MatrixClientPeg");
+
+export default class VoiceSettingsTab extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            mediaDevices: null,
+            activeAudioOutput: null,
+            activeAudioInput: null,
+            activeVideoInput: null,
+        };
+    }
+
+    componentWillMount(): void {
+        this._refreshMediaDevices();
+    }
+
+    _refreshMediaDevices = async () => {
+        this.setState({
+            mediaDevices: await CallMediaHandler.getDevices(),
+            activeAudioOutput: CallMediaHandler.getAudioOutput(),
+            activeAudioInput: CallMediaHandler.getAudioInput(),
+            activeVideoInput: CallMediaHandler.getVideoInput(),
+        });
+    };
+
+    _requestMediaPermissions = () => {
+        const getUserMedia = (
+            window.navigator.getUserMedia || window.navigator.webkitGetUserMedia || window.navigator.mozGetUserMedia
+        );
+        if (getUserMedia) {
+            return getUserMedia.apply(window.navigator, [
+                { video: true, audio: true },
+                this._refreshMediaDevices,
+                function() {
+                    const ErrorDialog = sdk.getComponent('dialogs.ErrorDialog');
+                    Modal.createTrackedDialog('No media permissions', '', ErrorDialog, {
+                        title: _t('No media permissions'),
+                        description: _t('You may need to manually permit Riot to access your microphone/webcam'),
+                    });
+                },
+            ]);
+        }
+    };
+
+    _setAudioOutput = (e) => {
+        CallMediaHandler.setAudioOutput(e.target.value);
+    };
+
+    _setAudioInput = (e) => {
+        CallMediaHandler.setAudioInput(e.target.value);
+    };
+
+    _setVideoInput = (e) => {
+        CallMediaHandler.setVideoInput(e.target.value);
+    };
+
+    _setForceTurn = (forced) => {
+        MatrixClientPeg.get().setForceTURN(forced);
+    };
+
+    _renderDeviceOptions(devices, category) {
+        return devices.map((d) => <option key={`${category}-${d.deviceId}`} value={d.deviceId}>{d.label}</option>);
+    }
+
+    render() {
+        const SettingsFlag = sdk.getComponent("views.elements.SettingsFlag");
+
+        let requestButton = null;
+        let speakerDropdown = null;
+        let microphoneDropdown = null;
+        let webcamDropdown = null;
+        if (this.state.mediaDevices === false) {
+            requestButton = (
+                <div>
+                    <p>{_t("Missing media permissions, click the button below to request.")}</p>
+                    <AccessibleButton onClick={this._requestMediaPermissions} kind="primary">
```

Comment:
This button looks unusually tall in the screenshot, is it correct?

Also, I think some margin below the button would look a bit better, so it's not squished into the toggles below.

Reply:
+1 to margin

and yes, the button is 40px instead of the intended 32px for some reason... I'll go make it, and all the others that are 40px, 32px.